### PR TITLE
[SPARK-10152] [SQL] Support Init script for hive-thriftserver

### PR DESCRIPTION
**Pull request by navis**
If some queries can be executed on thrift server in initialization stage(mostly for registering functions or macros), things are done much easier.

### CODE REVIEW

**Comment by andrewor14** [comment]
@yhuai

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala [hunk 1]
```diff
@@ -67,6 +68,46 @@ object HiveThriftServer2 extends Logging {
...
+  def runScript(sqlContext: HiveContext) {
+    val hiveConf = sqlContext.hiveconf
+    val script = hiveConf.get("hive.server2.init.script", null)
+    if (script != null && !script.isEmpty) {
+      getCommands(script).foreach { c =>
+        try {
+          LOG.warn("Running.. " + c)
+          sqlContext.runSqlHive(c).foreach { o => LOG.warn(o) }
+        } catch {
+          case t: Throwable => LOG.warn("Failed to run command " + c, t)
```
Comment (JoshRosen) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala]: This should probably catch `NonFatal` instead of throwable, no?

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala [hunk 2]
```diff
@@ -67,6 +68,46 @@ object HiveThriftServer2 extends Logging {
...
+        }
+      }
+    }
+  }
+
+  private def getCommands(input: String): Seq[String] = {
+    LOG.warn("Reading script " + input)
```
Comment (JoshRosen) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala]: Is this really a warning-level log message? This seems more like INFO to me.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala [hunk 3]
```diff
@@ -67,6 +68,46 @@ object HiveThriftServer2 extends Logging {
     }
   }
 
+  def runScript(sqlContext: HiveContext) {
+    val hiveConf = sqlContext.hiveconf
+    val script = hiveConf.get("hive.server2.init.script", null)
+    if (script != null && !script.isEmpty) {
+      getCommands(script).foreach { c =>
+        try {
+          LOG.warn("Running.. " + c)
```
Comment (JoshRosen) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala]: Same here: INFO log level?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala [hunk 4]
```diff
@@ -549,9 +549,11 @@ class HiveContext(sc: SparkContext) extends SQLContext(sc) with Logging {
   }
 
   protected[hive] def runSqlHive(sql: String): Seq[String] = {
-    if (sql.toLowerCase.contains("create temporary function")) {
+    val normalized = sql.trim.toLowerCase
+    if (normalized.contains("create temporary function") ||
+      normalized.contains("create temporary macro")) {
```
Comment (JoshRosen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala]: We don't actually support macros right now, right? I'd prefer to revert this change and handle it separately as part of your macro PR.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala [hunk 5]
```diff
@@ -67,6 +68,46 @@ object HiveThriftServer2 extends Logging {
     }
   }
 
+  def runScript(sqlContext: HiveContext) {
+    val hiveConf = sqlContext.hiveconf
+    val script = hiveConf.get("hive.server2.init.script", null)
```
Comment (JoshRosen) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala]: Is this a standard conf? Googling it didn't turn up anything.

Comment (JoshRosen) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala]: I'm not super-familiar with this part of Hive, but a bit of sleuthing suggests that there's a `hive.server2.global.init.file.location` configuration for init scripts: https://cwiki.apache.org/confluence/display/Hive/Setting+Up+HiveServer2#SettingUpHiveServer2-OptionalGlobalInitFile

Why not use that?

**Comment by JoshRosen** [comment]
If the Hive thriftserver supports init scripts, then I'm okay with doing the same, but it would be good to both clarify the configuration-related questions, add documentation, and add a unit test.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala (hunk 5 above)
Comment (chenghao-intel) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala]: Actually Spark SQL can load the init scripts, and the default value of `hive.server2.global.init.file.location` is $HIVE_CONF_DIR, not $SPARK_CONF_DIR or the current classpath. 

So this PR makes the same thing which is already supported, probably what we want is changing the default value to $SPARK_CONF_DIR?

Comment (chenghao-intel) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala]: After I created the env variable `HIVE_CONF_DIR` and place the file `.hiverc` under that folder, I met another exception.
```
org.apache.hive.service.cli.HiveSQLException: Failed to open new session: java.lang.RuntimeException: java.util.NoSuchElementException: key not found: SessionHandle [5a93b7b6-b011-49a4-a7c1-807bceac9800]
	at org.apache.hive.service.cli.session.SessionManager.openSession(SessionManager.java:266)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLSessionManager.openSession(SparkSQLSessionManager.scala:64)
	at org.apache.hive.service.cli.CLIService.openSessionWithImpersonation(CLIService.java:202)
	at org.apache.hive.service.cli.thrift.ThriftCLIService.getSessionHandle(ThriftCLIService.java:402)
	at org.apache.hive.service.cli.thrift.ThriftCLIService.OpenSession(ThriftCLIService.java:297)
	at org.apache.hive.service.cli.thrift.TCLIService$Processor$OpenSession.getResult(TCLIService.java:1253)
	at org.apache.hive.service.cli.thrift.TCLIService$Processor$OpenSession.getResult(TCLIService.java:1238)
	at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39)
	at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39)
	at org.apache.hive.service.auth.TSetIpAddressProcessor.process(TSetIpAddressProcessor.java:56)
	at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:285)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603)
	at java.lang.Thread.run(Thread.java:722)
Caused by: java.lang.RuntimeException: java.util.NoSuchElementException: key not found: SessionHandle [5a93b7b6-b011-49a4-a7c1-807bceac9800]
	at org.apache.hive.service.cli.session.HiveSessionProxy.invoke(HiveSessionProxy.java:83)
	at org.apache.hive.service.cli.session.HiveSessionProxy.access$000(HiveSessionProxy.java:36)
	at org.apache.hive.service.cli.session.HiveSessionProxy$1.run(HiveSessionProxy.java:63)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:415)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1548)
	at org.apache.hive.service.cli.session.HiveSessionProxy.invoke(HiveSessionProxy.java:59)
	at $Proxy24.open(Unknown Source)
	at org.apache.hive.service.cli.session.SessionManager.openSession(SessionManager.java:258)
	... 13 more
Caused by: java.util.NoSuchElementException: key not found: SessionHandle [5a93b7b6-b011-49a4-a7c1-807bceac9800]
	at scala.collection.MapLike$class.default(MapLike.scala:228)
	at scala.collection.AbstractMap.default(Map.scala:58)
	at scala.collection.mutable.HashMap.apply(HashMap.scala:64)
	at org.apache.spark.sql.hive.thriftserver.server.SparkSQLOperationManager.newExecuteStatementOperation(SparkSQLOperationManager.scala:47)
	at org.apache.hive.service.cli.session.HiveSessionImpl.executeStatementInternal(HiveSessionImpl.java:384)
	at org.apache.hive.service.cli.session.HiveSessionImpl.access$000(HiveSessionImpl.java:78)
	at org.apache.hive.service.cli.session.HiveSessionImpl$GlobalHivercFileProcessor.processCmd(HiveSessionImpl.java:172)
	at org.apache.hadoop.hive.common.cli.HiveFileProcessor.processLine(HiveFileProcessor.java:87)
	at org.apache.hadoop.hive.common.cli.HiveFileProcessor.processReader(HiveFileProcessor.java:66)
	at org.apache.hadoop.hive.common.cli.HiveFileProcessor.processFile(HiveFileProcessor.java:37)
	at org.apache.hive.service.cli.session.HiveSessionImpl.processGlobalInitFile(HiveSessionImpl.java:193)
	at org.apache.hive.service.cli.session.HiveSessionImpl.open(HiveSessionImpl.java:146)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:601)
	at org.apache.hive.service.cli.session.HiveSessionProxy.invoke(HiveSessionProxy.java:78)
	... 21 more
```

Maybe we can solve that in this PR.

Comment (vincentye38) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala]: Do you resolve this HiveSQLException? I ran into the same exception.
